### ISSUE TITLE
Missing values for non-static final fields

### ISSUE THREAD

**Issue by rankinc** [classgraph/classgraph#379]
Hi,

I have been trying to update a (_very!_) old code-base from FastClasspathScanner 2.18.1 to ClassGraph 4.8.52. This code must be able to scan Kotlin as well as Java classes, and while I have replicated most of the functionality, I am still struggling to persuade ClassGraph to handle this:
```kotlin
class HasJvmField {
    @JvmField
    val stringValue = "Hello World"
}
```
The problem is that ClassGraph doesn't realise that this field has an initialier, whereas FastClasspathScanner did. The Kotlin compiler generates the following byte-code for this case:
```
  public final java.lang.String stringValue;
    descriptor: Ljava/lang/String;
    flags: ACC_PUBLIC, ACC_FINAL
    ConstantValue: String Hello World
    RuntimeInvisibleAnnotations:
      0: #9()
      1: #10()

   #9 = Utf8               Lkotlin/jvm/JvmField;
  #10 = Utf8               Lorg/jetbrains/annotations/NotNull;
```
I suspect this is because `stringValue` is **not** a `static` field and so enabling
```java
.enableStaticFinalFieldConstantInitializerValues()
```
does not help me here.

Is there a workaround for this please?

**Comment by lukehutch** [comment]
Interesting, I think Java doesn't assign constant field initializer values unless the field is static, but the classfile format doesn't preclude it. I'll drop the static (and the final) requirement, and will release a new version soon.

**Comment by lukehutch** [comment]
Also you're the first user I have heard of that actually uses this unique feature! It's one of the things I first needed when I wrote FastClasspathScanner.

**Comment by rankinc** [comment]
> Interesting, I think Java doesn't assign constant field initializer values unless the field is static, but the classfile format doesn't preclude it. I'll drop the static (and the final) requirement, and will release a new version soon.

There's no accounting for how the Kotlin compiler writers will interpret the Java classfile format :wink:. Despite the constant initializer, the Kotlin constructor does set the field value too:
```
    descriptor: ()V
    flags: ACC_PUBLIC
    Code:
      stack=2, locals=1, args_size=1
         0: aload_0
         1: invokespecial #14                 // Method java/lang/Object."<init>":()V
         4: aload_0
         5: ldc           #8                  // String Hello World
         7: putfield      #16                 // Field stringValue:Ljava/lang/String;
        10: return
```
I was envisaging a brand new ClassGraph API like:
```java
.enableFieldConstantInitializerValues()
```
which performs an implicit
```java
.enableStaticFinalFieldConstantInitializerValues()
```

**Comment by lukehutch** [comment]
Well as you discovered, there's no need to check the field modifiers at all... I'll just create the new call you suggest, which will not check modifiers, and deprecate the old one.

**Comment by lukehutch** [comment]
@rankinc I removed the check for `static final`. Please test the git master version, and let me know if it works for you.

**Comment by lukehutch** [comment]
@rankinc Turns out that the Java compiler definitely doesn't store constant initializer values in the constant pool for a non-static field. Actually the reason why the Kotlin compiler manually sets the value in the constructor is because the JVM ignores constant initializer values for non-static fields, even though the Kotlin compiler puts the value there:

https://docs.oracle.com/javase/specs/jvms/se11/html/jvms-4.html#jvms-4.7.2

> `A ConstantValue` attribute represents the value of a constant expression (JLS §15.28), and is used as follows:
> 
> If the `ACC_STATIC` flag ... is set, then the field ... is assigned the value represented by its `ConstantValue` attribute as part of the initialization of the class or interface declaring the field (§5.5). This occurs prior to the invocation of the class or interface initialization method of that class or interface (§2.9.2).
> 
> Otherwise, the Java Virtual Machine must silently ignore the attribute.

So actually Kotlin is broken here, and the fact that the code worked under FastClasspathScanner is an accident. Even if I ignore the `static` modifier on fields (which would fix your code), it is possible that a future version of the Kotlin compiler will change its code generation to drop this unnecessary field attribute, which is currently ignored by the JVM.

I was also checking for `final` though, and it turns out that is unnecessary, so I should at least drop that check.

I'm thinking it won't hurt to simply remove both of the `static final` checks, but you should not rely on this working forever.

**Comment by rankinc** [comment]
Thanks, this fixes the problem. Migrating our tool's output so that we don't  need the field values will be a separate problem, unfortunately.

**Comment by lukehutch** [comment]
The checkin fixes the problem, sure -- but asked in a different way -- should the method name even be changed, or the `static final` checks be dropped, if Kotlin doesn't actually conform to the spec (with respect to `static`), or to the behavior of `javac` (with respect to `final`)?

**Comment by lukehutch** [comment]
Is there a way for you to change the fields to `static final` for this particular usecase? Or do you rely on this pervasively, for more than one thing? And do they have to be non-`static` and/or non-`final`?

**Comment by lukehutch** [comment]
One "meet in the middle" approach is to simply drop the `static final` check, but leave the method name referencing `StaticFinal`, and just state in the Javadoc that it might work for non-`static` / non-`final` fields in Kotlin (although may break in the future), but only for `static final` fields in Java. I think that's what I will go with, since it doesn't negatively impact any Java user, and doesn't require an API change, but still gets you what you need.

**Comment by rankinc** [comment]
We have a public API / ABI, and a public commitment not to break it. So it's going to take some negotiation...

**Comment by lukehutch** [comment]
It's OK, you won't need the negotiation, see my previous comment. I'll simply drop the `static final` check and keep the ClassGraph method name the same. About to check in code that does that. Your only risk moving forward is that the Kotlin compiler might simply stop generating constant initializers for non-static fields, in order to conform more closely to the spec. (Generating the value doesn't hurt, other than adding a few bytes to the classfile, but the value is ignored by the JVM.) You should probably ask the Kotlin compiler developers if they even care about this. If they don't, maybe this will work forever. If they do care about improving conformance with the classfile spec and `javac`'s own behavior, then maybe you need to look at breaking the API/ABI in future. (Incidentally this would not be the first time I have had to work around Kotlin's habit of generating slightly wrong classfiles...)

**Comment by lukehutch** [comment]
Released in 4.8.53. Just use `.enableStaticFinalFieldConstantInitializerValues()` and ignore the `StaticFinal` part of the name. Thanks for the report!
